lang/csharp/src/apache/test/Reflect/TestFromAvroProject.cs

```diff
@@ -0,0 +1,252 @@
+
+                var avroWriter = new ReflectWriter<Z>(schema);
+                var avroReader = new ReflectReader<Z>(schema, schema);
+
+                byte[] serialized;
+
+                using (var stream = new MemoryStream(256))
+                {
+                    avroWriter.Write(z, new BinaryEncoder(stream));
+                    serialized = stream.ToArray();
+                }
+
+                Z deserialized = null;
+                using (var stream = new MemoryStream(serialized))
+                {
+                    deserialized = avroReader.Read(default(Z), new BinaryDecoder(stream));
+                }
+
+                return deserialized;
+            }
+            catch (Exception ex)
+            {
+                Console.WriteLine(ex.ToString());
+                throw ex;
+            }
+        }
+
+        [TestCase]
+        public void DefaultZ()
+        {
+            var z = new Z()
+            {
+                myBytes = new byte[10],
+                myString = "123",
+                myFixed = new byte[16],
+                myA = new A(),
+                myArray = new List<byte[]>(),
+                myArray2 = new List<newRec>(),
+                myMap = new Dictionary<string, string>(),
+                myMap2 = new Dictionary<string, newRec>(),
+                myArray3 = new List<List<object>>()
+            };
+
+            var zz = SerializeDeserialize(z);
+        }
+
+        [TestCase]
+        public void PopulatedZ()
+        {
+            var z = new Z()
+            {
+                myUInt = 1,
+                myULong = 2L,
+                myUBool = true,
+                myUDouble = 3.14,
+                myUFloat = (float)1.59E-3,
+                myUBytes = new byte[3] { 0x01, 0x02, 0x03 },
+                myUString = "abc",
+                myInt = 1,
+                myLong = 2L,
+                myBool = true,
+                myDouble = 3.14,
+                myFloat = (float)1.59E-2,
+                myBytes = new byte[3] { 0x01, 0x02, 0x03 },
+                myString = "def",
+                myNull = null,
+                myFixed = new byte[16] { 0x01, 0x02, 0x03, 0x04, 0x01, 0x02, 0x03, 0x04, 0x01, 0x02, 0x03, 0x04, 0x01, 0x02, 0x03, 0x04 },
+                myA = new A() { f1 = 3L },
+                myE = MyEnum.B,
+                myArray = new List<byte[]>() { new byte[] { 0x01, 0x02, 0x03, 0x04 } },
+                myArray2 = new List<newRec>() { new newRec() { f1 = 4L } },
+                myMap = new Dictionary<string, string>()
+                {
+                    ["abc"] = "123"
+                },
+                myMap2 = new Dictionary<string, newRec>()
+                {
+                    ["abc"] = new newRec() { f1 = 5L }
+                },
+                myObject = new A() { f1 = 6L },
+                myArray3 = new List<List<object>>() { new List<object>() { 7.0, "def" } }
+            };
+
+            var zz = SerializeDeserialize(z);
```

Comment:
We should assert that all the properties of `zz` equal those of `z`.

Reply:
Done